File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/resolution/SemanticRetrievalPlanner.java
package com.foundgine.core.semantic.resolution;

import com.foundgine.core.semantic.SemanticField;

public final class SemanticRetrievalPlanner {
    private SemanticRetrievalPlanner() {}

    public static RetrievalStrategy select(SemanticField field) {
        return select(field, RetrievalStrategy.RELATIONAL);
    }

    /**
     * Selects a provider-neutral retrieval strategy. It never chooses a concrete
     * search/database provider.
     */
    public static RetrievalStrategy select(SemanticField field, RetrievalStrategy requested) {
        if (requested == RetrievalStrategy.RELATIONAL) return RetrievalStrategy.RELATIONAL;
        return switch (requested) {
            case FUZZY, FULL_TEXT, SEARCH, VECTOR -> requested;
            case GRAPH_SIMILARITY -> RetrievalStrategy.GRAPH_SIMILARITY;
            case RELATIONAL -> RetrievalStrategy.RELATIONAL;
        };
    }

    public static boolean requiresApproximateRetrieval(RetrievalStrategy strategy) {
        return switch (strategy) {
            case FULL_TEXT, SEARCH, FUZZY, VECTOR, GRAPH_SIMILARITY -> true;
            case RELATIONAL -> false;
        };
    }
}

File: src/csharp/Foundgine.Core/Semantic/Resolution/SemanticRetrievalPlanner.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Resolution;

/// <summary>
/// Chooses whether a predicate can remain relational or requires approximate retrieval.
/// It deliberately does not choose a concrete provider.
/// </summary>
public static class SemanticRetrievalPlanner
{
    public static RetrievalStrategy Select(
        SemanticField field,
        RetrievalStrategy requested = RetrievalStrategy.Relational)
    {
        if (requested == RetrievalStrategy.Relational)
            return RetrievalStrategy.Relational;

        if (requested is RetrievalStrategy.Fuzzy or RetrievalStrategy.FullText or RetrievalStrategy.Search
            or RetrievalStrategy.Vector)
            return requested;

        return RetrievalStrategy.GraphSimilarity;
    }

    public static bool RequiresApproximateRetrieval(RetrievalStrategy strategy) =>
        strategy is RetrievalStrategy.FullText
            or RetrievalStrategy.Search
            or RetrievalStrategy.Fuzzy
            or RetrievalStrategy.Vector
            or RetrievalStrategy.GraphSimilarity;
}